inloop/statistics/views.py

```diff
@@ -0,0 +1,137 @@
+import json
+from collections import defaultdict
+from json import JSONDecodeError
+
+import django.db.models.functions as functions
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.db import models
+from django.http import JsonResponse
+from django.views import View
+from django.views.generic import TemplateView
+
+from inloop.solutions.models import Solution
+from inloop.tasks.models import Category, Task
+
+
+def get_truncator(
+    field_name: str, granularity: str
+) -> functions.datetime.TruncBase:
+    return {
+        "minute": functions.TruncMinute(field_name),
+        "hour": functions.TruncHour(field_name),
+        "day": functions.TruncDay(field_name),
+        "month": functions.TruncMonth(field_name),
+        "year": functions.TruncYear(field_name),
+    }[granularity]
+
+
+class FailureJsonResponse(JsonResponse):
+    reason = None
+
+    def __init__(self, *args, **kwargs):
+        if self.reason is None:
+            raise NotImplementedError("Failure JSON Responses must include a reason!")
+        super().__init__({
+            "reason": self.reason
+        }, *args, **kwargs)
+
+
+class IncorrectAccessMethod(FailureJsonResponse):
+    reason = "incorrect_access_method"
+    status_code = 405
+
+
+class MalformedJson(FailureJsonResponse):
+    reason = "malformed_json"
+    status_code = 400
+
+
+class AdminView(UserPassesTestMixin, LoginRequiredMixin, View):
+    def test_func(self):
+        if not self.request.user:
+            return False
+        return self.request.user.is_superuser or self.request.user.is_staff
+
+
+class StatisticsView(AdminView, TemplateView):
+    template_name = "statistics/index.html"
+
+
+class SolutionsHistogramView(AdminView, TemplateView):
+    template_name = "statistics/solutions_histogram.html"
+
+    def get_context_data(self):
+        return {
+            "categories": Category.objects.all()
+        }
+
+    def post(self, request) -> JsonResponse:
```

Comment:
The `GET` method of this view is already used to render the template. For me, this is ok, because I am essentially posting a json body with parameters, like if I was posting a form.

Reply:
In the course of refactoring, however, I will separate these paths into two distinct views.

Reply:
Which means, that I will also refactor this to `GET` as suggested.